### PR TITLE
Unit test and fix for bug (#154)

### PR DESCRIPTION
**Pull request by andrey-radchenko**

### CODE REVIEW

File: rxjava-reactive-streams/src/main/java/rx/internal/reactivestreams/SingleAsPublisher.java [hunk 1]
```diff
@@ -16,56 +16,55 @@
...
         this.single = single;
     }
 
     @Override
     public void subscribe(Subscriber<? super T> s) {
         SingleAsPublisherSubscriber<T> parent = new SingleAsPublisherSubscriber<T>(s);
-        s.onSubscribe(parent);
-        
         single.subscribe(parent);
+        s.onSubscribe(parent);
```
Comment (akarnokd) [inline rxjava-reactive-streams/src/main/java/rx/internal/reactivestreams/SingleAsPublisher.java:44]: This  violates the RS contract. OnSubscribe must happen before an other calls to onXXX.

File: rxjava-reactive-streams/src/test/java/rx/reactivestreams/test/PublisherAsSingleTest.java [hunk 2]
```diff
@@ -22,8 +26,10 @@ public void nullCheck() {
     @Test
     public void just() {
         TestSubscriber<Object> ts = new TestSubscriber<Object>();
-        
-        toSingle(toPublisher(Observable.just(1))).subscribe(ts);
+
```
Comment (akarnokd) [inline rxjava-reactive-streams/src/test/java/rx/reactivestreams/test/PublisherAsSingleTest.java:29]: Please don't change existing unit tests and add new ones instead.

**Review by akarnokd — CHANGES_REQUESTED** [review]

**Comment by akarnokd** [comment]
The bug is that there is a `return` missing just before [L90](https://github.com/ReactiveX/RxJavaReactiveStreams/blob/1.x/rxjava-reactive-streams/src/main/java/rx/internal/reactivestreams/SingleAsPublisher.java#L90)

File: rxjava-reactive-streams/src/test/java/rx/reactivestreams/test/PublisherAsSingleTest.java (hunk 2 above)
Comment (andrey-radchenko) [reply rxjava-reactive-streams/src/test/java/rx/reactivestreams/test/PublisherAsSingleTest.java:29]: I will restore original just() unit test and will add singleJust() that was hanging.

File: rxjava-reactive-streams/src/main/java/rx/internal/reactivestreams/SingleAsPublisher.java (hunk 1 above)
Comment (andrey-radchenko) [reply rxjava-reactive-streams/src/main/java/rx/internal/reactivestreams/SingleAsPublisher.java:44]: Will restore origin subscribe() flow

**Comment by akarnokd** [comment]
Maybe its better if you start over from a fresh branch of main.

**Comment by andrey-radchenko** [comment]
Well, I started over from the tip of 1.x and added the unit test. However, adding `return` just before [L90](https://github.com/ReactiveX/RxJavaReactiveStreams/blob/1.x/rxjava-reactive-streams/src/main/java/rx/internal/reactivestreams/SingleAsPublisher.java#L90) has not changed anything. The problem is with the [infinite loop](https://github.com/ReactiveX/RxJavaReactiveStreams/blob/1.x/rxjava-reactive-streams/src/main/java/rx/internal/reactivestreams/SingleAsPublisher.java#L111-L128). Without the actual `single` subscription, the current state will not change from outside the loop and there is no transition from NO_REQUEST_NO_DATA `state` within the loop, so that will keep it spinning forever.

**Comment by akarnokd** [comment]
I see. Try this:

``` java

        @Override
        public void onSuccess(T value) {
            if (cancelled) {
                return;
            }
            if (value == null) {
                state.lazySet(HAS_REQUEST_HAS_VALUE);
                actual.onError(new NullPointerException("value"));
                return;
            }
            for (;;) {
                int s = state.get();

                if (s == NO_REQUEST_HAS_VALUE || s == HAS_REQUEST_HAS_VALUE || cancelled) {
                    break;
                } else
                if (s == HAS_REQUEST_NO_VALUE) {
                    state.lazySet(HAS_REQUEST_HAS_VALUE);
                    actual.onNext(value);
                    if (!cancelled) {
                        actual.onComplete();
                    }
                    break;
                } else {
                    this.value = value;
                    if (state.compareAndSet(s, NO_REQUEST_HAS_VALUE)) {
                        break;
                    }
                }
            }
        }

        @Override
        public void onError(Throwable error) {
            if (cancelled) {
                return;
            }
            state.lazySet(HAS_REQUEST_HAS_VALUE);
            actual.onError(error);
        }

        @Override
        public void request(long n) {
            if (n > 0) {
                for (;;) {
                    int s = state.get();
                    if (s == HAS_REQUEST_HAS_VALUE || s == HAS_REQUEST_NO_VALUE || cancelled) {
                        break;
                    } else
                    if (s == NO_REQUEST_HAS_VALUE) {
                        if (state.compareAndSet(s, HAS_REQUEST_HAS_VALUE)) {
                            T v = value;
                            value = null;

                            actual.onNext(v);
                            if (!cancelled) {
                                actual.onComplete();
                            }
                        }
                        break;
                    }
                    if (state.compareAndSet(NO_REQUEST_NO_VALUE, HAS_REQUEST_NO_VALUE)) {
                        break;
                    }
                }
            }
        }
```

**Comment by andrey-radchenko** [comment]
Yes, this works. Breaking the request loop on `NO_REQUEST_NO_VALUE` was missing there. Committed the change and created PR. (#156)

Please decline the previous PR. (#155)

**Comment by akarnokd** [comment]
Closing via #156.

**Pull request closed without merging**